localstack/aws/chain.py

```diff
@@ -0,0 +1,140 @@
+"""
+The core concepts of the HandlerChain.
+"""
+import logging
+from typing import Any, Callable, List, Optional
+
+from localstack.http import Response
+
+from .api import RequestContext
+
+LOG = logging.getLogger(__name__)
+
+Handler = Callable[["HandlerChain", RequestContext, Response], None]
+ExceptionHandler = Callable[["HandlerChain", Exception, RequestContext, Response], None]
+
+
+class HandlerChain:
+    # handlers
+    request_handlers: List[Handler]
+    response_handlers: List[Handler]
+    exception_handlers: List[ExceptionHandler]
+
+    # behavior configuration
+    stop_on_error: bool = True
+    raise_on_error: bool = False
+
+    # state
+    stopped: bool
+    terminated: bool
+    error: Optional[Exception]
+    response: Optional[Response]
+    context: Optional[RequestContext]
+
+    def __init__(
+        self,
+        request_handlers: List[Handler] = None,
+        response_handlers: List[Handler] = None,
+        exception_handlers: List[ExceptionHandler] = None,
+    ) -> None:
+        super().__init__()
+        self.request_handlers = request_handlers or list()
+        self.response_handlers = response_handlers or list()
+        self.exception_handlers = exception_handlers or list()
+
+        self.stopped = False
+        self.terminated = False
+        self.error = None
+        self.response = None
+        self.context = None
+
+    def handle(self, context: RequestContext, response: Response):
+        self.context = context
+        self.response = response
+
+        for handler in self.request_handlers:
+            try:
+                handler(self, self.context, response)
+            except Exception as e:
+                # prepare the continuation behavior, but exception handlers could overwrite it
+                if self.raise_on_error:
+                    self.error = e
+                if self.stop_on_error:
+                    self.stopped = True
+
+                # call exception handlers safely
+                self._call_exception_handlers(e, response)
+
+            # decide next step
+            if self.error:
+                raise self.error
```

Comment:
Over time we could consider introducing some sort of a finally-handler to reliably perform cleanup logic. If we raise the error here, then the response handlers would not get called in line 77, and the request context in `quart.globals._request_ctx_stack` would potentially remain on the stack. 

Should not be an issue for now (looks like none of the handlers are calling `.throw()` yet, and since we have `raise_on_error=False` by default), but something we can consider for the future iterations. 👍

Reply:
yea this is also a construct that was really introduced in the interest of generality. definitely not something that you would use in a normal production setting or with the `LocalstackGateway`. but it may come in handy for testing/debugging. other handler chains may want to re-raise errors immediately. :shrug: let's see whether we really need it

Reply:
👌